coverage/src/test/kotlin/org/rust/coverage/RsCoverageTest.kt

```diff
@@ -0,0 +1,222 @@
+                    }
+                }                                   // Hits: 1
+            """)
+        }
+    }
+
+    fun `test for function call`() = doTest {
+        toml("Cargo.toml", """
+            [package]
+            name = "hello"
+            version = "0.1.0"
+            authors = []
+        """)
+
+        dir("src") {
+            rust("lib.rs", """
+                fn foo() {                      // Hits: 4
+                    println!("Hello, world!");  // Hits: 4
+                }                               // Hits: 4
+                
+                fn main() {                     // Hits: 1
+                    for _ in 1..5 {             // Hits: 5
+                        foo();                  // Hits: 4
+                    }
+                }                               // Hits: 1
+            """)
+        }
+    }
+
+    fun `test tests`() = doTest(runTests = true) {
+        toml("Cargo.toml", """
+            [package]
+            name = "hello"
+            version = "0.1.0"
+            authors = []
+        """)
+
+        dir("src") {
+            rust("lib.rs", """
+                fn foo() {                      // Hits: 2
+                    println!("Hello, world!");  // Hits: 2
+                }                               // Hits: 2
+                
+                #[test]
+                fn test1() {                    // Hits: 2
+                    foo();                      // Hits: 1
+                }                               // Hits: 2
+                
+                #[test]
+                fn test2() {                    // Hits: 2
+                    foo();                      // Hits: 1
+                }                               // Hits: 2
+            """)
+        }
+    }
+
+    private fun executeGrcov(configuration: RunConfiguration) {
+        val executor = DefaultRunExecutor.getRunExecutorInstance()
+        val environment = ExecutionEnvironmentBuilder.create(executor, configuration).build()
+        val state = environment.state as CargoRunStateBase
+        GrcovRunner().execute(environment, null, state)
+    }
+
+    private fun doTest(runTests: Boolean = false, builder: FileTreeBuilder.() -> Unit) {
+        buildProject(builder)
+
+        val expected = hashMapOf<String, List<Pair<Int, Int>>>()
+        VfsUtil.iterateChildrenRecursively(cargoProjectDirectory, null) { fileOrDir ->
+            if (!fileOrDir.isRustFile) return@iterateChildrenRecursively true
+            val psiFile = fileOrDir.toPsiFile(project) ?: return@iterateChildrenRecursively true
+            expected[fileOrDir.path] = hitsFrom(psiFile.text)
+            true
+        }
+
+        val configuration = createConfiguration()
+        if (runTests) {
+            configuration.command = "test"
+        }
+
+        executeGrcov(configuration)
+        runWithInvocationEventsDispatching("Failed to fetch coverage data") { lastProjectData != null }
```

Comment:
JFYI: there is `CoverageSuiteListener` and probably we can use `com.intellij.coverage.CoverageSuiteListener#afterSuiteChosen` to be notified that running is finished

Reply:
I use `runWithInvocationEventsDispatching` because it calls `UIUtil.dispatchAllInvocationEvents` periodically.